archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java

```diff
@@ -0,0 +1,147 @@
+package com.netflix.config.source;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Splitter;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Maps;
+import com.netflix.config.WatchedConfigurationSource;
+import com.netflix.config.WatchedUpdateListener;
+import com.netflix.config.WatchedUpdateResult;
+import org.boon.core.Handler;
+import org.boon.etcd.Etcd;
+import org.boon.etcd.Node;
+import org.boon.etcd.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+
+/**
+ * Implementation of the dynamic {@link WatchedConfigurationSource} for Etcd
+ *
+ * This implementation requires the path to the Etcd directory that contains
+ * nodes that represent each managed configuration property.
+ * 
+ * An example Etcd configuration path is /<my-app>/config
+ * An example Etcd property node path is /<my-app>/config/com.fluxcapacitor.my.property
+ * 
+ * When a property is mutated via Etcd a callback will be notified and the value managed
+ * by EtcdConfigurationSource will be updated. Similar to other dynamic configuration
+ * source (ie. DynamoDB, etc.)
+ * 
+ * @author spoon16
+ */
+public class EtcdConfigurationSource implements WatchedConfigurationSource {
+    private static final Logger logger = LoggerFactory.getLogger(EtcdConfigurationSource.class);
+    private static final Splitter keySplitter = Splitter.on('/');
+
+    private final Map<String, Object> valueCache = Maps.newConcurrentMap();
+    private final List<WatchedUpdateListener> listeners = new CopyOnWriteArrayList<WatchedUpdateListener>();
+
+    private final Etcd etcd;
+    private final String configPath;
+
+    private Handler<Response> updateHandler = new Handler<Response>() {
+        @Override
+        public void handle(Response updateResponse) {
+            if (updateResponse.wasError()) {
+                logger.error("Etcd failed with an error response: %s", updateResponse);
+            }
+
+            final Map<String, Object> create = Maps.newHashMap();
+            final Map<String, Object> set = Maps.newHashMap();
+            final Map<String, Object> delete = Maps.newHashMap();
+
+            final String action = updateResponse.action().toLowerCase();
```

Comment:
only if I duplicate the etcd.waitRecursive call on l:89

It'd probably be best to capture the wasError condition, throw an exception and have a finally that calls etcd.waitRecursive.  But I'm not sure that really makes things more readable.  I'll just duplicate l:89 and return early.